File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    
    int n = (int)triangle.size();
    vector<int> prev(n);
    vector<int> curr(n);
    
    for(int i=0; i<n; i++){
        for(int j=0; j<i+1; j++){
            if(j==0)
                curr[j] = prev[0] + triangle[i][j];
            else if(i==j)
                curr[j] = prev[j-1] + triangle[i][j];
            else
                curr[j] = max(prev[j-1], prev[j]) + triangle[i][j];
        }
        prev= curr;
    }
    
    answer = *max_element(prev.begin(), prev.end());
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    answer = 0
    n = len(triangle)
    prev = [0] * n
    curr = [0] * n
    
    for i in range(n):
        for j in range(i+1):
            if j==0:
                curr[j] = prev[0] + triangle[i][j]
            elif j==i:
                curr[j] = prev[j-1] + triangle[i][j]
            else:
                curr[j] = max(prev[j-1],prev[j]) + triangle[i][j]
                
        for i in range(n):
            prev[i] = curr[i]
        
    answer = max(prev)
    return answer